File: packages/backend-shim-c/src/domains/geometry.c
#include "tspice_backend_shim.h"

#include "SpiceUsr.h"

#include <string.h>

int tspice_subpnt(
    const char *method,
    const char *target,
    double et,
    const char *fixref,
    const char *abcorr,
    const char *observer,
    double *outSpoint3,
    double *outTrgepc,
    double *outSrfvec3,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outTrgepc) {
    *outTrgepc = 0.0;
  }

  SpiceDouble spoint[3];
  SpiceDouble trgepc = 0.0;
  SpiceDouble srfvec[3];
  subpnt_c(method, target, (SpiceDouble)et, fixref, abcorr, observer, spoint, &trgepc, srfvec);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outSpoint3) {
    outSpoint3[0] = (double)spoint[0];
    outSpoint3[1] = (double)spoint[1];
    outSpoint3[2] = (double)spoint[2];
  }
  if (outTrgepc) {
    *outTrgepc = (double)trgepc;
  }
  if (outSrfvec3) {
    outSrfvec3[0] = (double)srfvec[0];
    outSrfvec3[1] = (double)srfvec[1];
    outSrfvec3[2] = (double)srfvec[2];
  }

  return 0;
}

int tspice_subslr(
    const char *method,
    const char *target,
    double et,
    const char *fixref,
    const char *abcorr,
    const char *observer,
    double *outSpoint3,
    double *outTrgepc,
    double *outSrfvec3,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outTrgepc) {
    *outTrgepc = 0.0;
  }

  SpiceDouble spoint[3];
  SpiceDouble trgepc = 0.0;
  SpiceDouble srfvec[3];
  subslr_c(method, target, (SpiceDouble)et, fixref, abcorr, observer, spoint, &trgepc, srfvec);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outSpoint3) {
    outSpoint3[0] = (double)spoint[0];
    outSpoint3[1] = (double)spoint[1];
    outSpoint3[2] = (double)spoint[2];
  }
  if (outTrgepc) {
    *outTrgepc = (double)trgepc;
  }
  if (outSrfvec3) {
    outSrfvec3[0] = (double)srfvec[0];
    outSrfvec3[1] = (double)srfvec[1];
    outSrfvec3[2] = (double)srfvec[2];
  }

  return 0;
}

int tspice_sincpt(
    const char *method,
    const char *target,
    double et,
    const char *fixref,
    const char *abcorr,
    const char *observer,
    const char *dref,
    const double *dvec3,
    double *outSpoint3,
    double *outTrgepc,
    double *outSrfvec3,
    int *outFound,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outTrgepc) {
    *outTrgepc = 0.0;
  }
  if (outFound) {
    *outFound = 0;
  }

  SpiceDouble dvec[3];
  dvec[0] = dvec3 ? (SpiceDouble)dvec3[0] : 0.0;
  dvec[1] = dvec3 ? (SpiceDouble)dvec3[1] : 0.0;
  dvec[2] = dvec3 ? (SpiceDouble)dvec3[2] : 0.0;

  SpiceDouble spoint[3];
  SpiceDouble trgepc = 0.0;
  SpiceDouble srfvec[3];
  SpiceBoolean found = SPICEFALSE;

  sincpt_c(method, target, (SpiceDouble)et, fixref, abcorr, observer, dref, dvec, spoint, &trgepc, srfvec, &found);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outFound) {
    *outFound = found == SPICETRUE ? 1 : 0;
  }

  if (found == SPICETRUE) {
    if (outSpoint3) {
      outSpoint3[0] = (double)spoint[0];
      outSpoint3[1] = (double)spoint[1];
      outSpoint3[2] = (double)spoint[2];
    }
    if (outTrgepc) {
      *outTrgepc = (double)trgepc;
    }
    if (outSrfvec3) {
      outSrfvec3[0] = (double)srfvec[0];
      outSrfvec3[1] = (double)srfvec[1];
      outSrfvec3[2] = (double)srfvec[2];
    }
  }

  return 0;
}

int tspice_ilumin(
    const char *method,
    const char *target,
    double et,
    const char *fixref,
    const char *abcorr,
    const char *observer,
    const double *spoint3,
    double *outTrgepc,
    double *outSrfvec3,
    double *outPhase,
    double *outIncdnc,
    double *outEmissn,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outTrgepc) {
    *outTrgepc = 0.0;
  }
  if (outPhase) {
    *outPhase = 0.0;
  }
  if (outIncdnc) {
    *outIncdnc = 0.0;
  }
  if (outEmissn) {
    *outEmissn = 0.0;
  }

  SpiceDouble spoint[3];
  spoint[0] = spoint3 ? (SpiceDouble)spoint3[0] : 0.0;
  spoint[1] = spoint3 ? (SpiceDouble)spoint3[1] : 0.0;
  spoint[2] = spoint3 ? (SpiceDouble)spoint3[2] : 0.0;

  SpiceDouble trgepc = 0.0;
  SpiceDouble srfvec[3];
  SpiceDouble phase = 0.0;
  SpiceDouble incdnc = 0.0;
  SpiceDouble emissn = 0.0;

  ilumin_c(method, target, (SpiceDouble)et, fixref, abcorr, observer, spoint, &trgepc, srfvec, &phase, &incdnc, &emissn);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outTrgepc) {
    *outTrgepc = (double)trgepc;
  }
  if (outSrfvec3) {
    outSrfvec3[0] = (double)srfvec[0];
    outSrfvec3[1] = (double)srfvec[1];
    outSrfvec3[2] = (double)srfvec[2];
  }
  if (outPhase) {
    *outPhase = (double)phase;
  }
  if (outIncdnc) {
    *outIncdnc = (double)incdnc;
  }
  if (outEmissn) {
    *outEmissn = (double)emissn;
  }

  return 0;
}

int tspice_illumg(
    const char *method,
    const char *target,
    const char *ilusrc,
    double et,
    const char *fixref,
    const char *abcorr,
    const char *obsrvr,
    const double *spoint3,
    double *outTrgepc,
    double *outSrfvec3,
    double *outPhase,
    double *outIncdnc,
    double *outEmissn,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outTrgepc) {
    *outTrgepc = 0.0;
  }
  if (outPhase) {
    *outPhase = 0.0;
  }
  if (outIncdnc) {
    *outIncdnc = 0.0;
  }
  if (outEmissn) {
    *outEmissn = 0.0;
  }

  SpiceDouble spoint[3];
  spoint[0] = spoint3 ? (SpiceDouble)spoint3[0] : 0.0;
  spoint[1] = spoint3 ? (SpiceDouble)spoint3[1] : 0.0;
  spoint[2] = spoint3 ? (SpiceDouble)spoint3[2] : 0.0;

  SpiceDouble trgepc = 0.0;
  SpiceDouble srfvec[3];
  SpiceDouble phase = 0.0;
  SpiceDouble incdnc = 0.0;
  SpiceDouble emissn = 0.0;

  illumg_c(
      method,
      target,
      ilusrc,
      (SpiceDouble)et,
      fixref,
      abcorr,
      obsrvr,
      spoint,
      &trgepc,
      srfvec,
      &phase,
      &incdnc,
      &emissn);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outTrgepc) {
    *outTrgepc = (double)trgepc;
  }
  if (outSrfvec3) {
    outSrfvec3[0] = (double)srfvec[0];
    outSrfvec3[1] = (double)srfvec[1];
    outSrfvec3[2] = (double)srfvec[2];
  }
  if (outPhase) {
    *outPhase = (double)phase;
  }
  if (outIncdnc) {
    *outIncdnc = (double)incdnc;
  }
  if (outEmissn) {
    *outEmissn = (double)emissn;
  }

  return 0;
}

int tspice_illumf(
    const char *method,
    const char *target,
    const char *ilusrc,
    double et,
    const char *fixref,
    const char *abcorr,
    const char *obsrvr,
    const double *spoint3,
    double *outTrgepc,
    double *outSrfvec3,
    double *outPhase,
    double *outIncdnc,
    double *outEmissn,
    int *outVisibl,
    int *outLit,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outTrgepc) {
    *outTrgepc = 0.0;
  }
  if (outPhase) {
    *outPhase = 0.0;
  }
  if (outIncdnc) {
    *outIncdnc = 0.0;
  }
  if (outEmissn) {
    *outEmissn = 0.0;
  }
  if (outVisibl) {
    *outVisibl = 0;
  }
  if (outLit) {
    *outLit = 0;
  }

  SpiceDouble spoint[3];
  spoint[0] = spoint3 ? (SpiceDouble)spoint3[0] : 0.0;
  spoint[1] = spoint3 ? (SpiceDouble)spoint3[1] : 0.0;
  spoint[2] = spoint3 ? (SpiceDouble)spoint3[2] : 0.0;

  SpiceDouble trgepc = 0.0;
  SpiceDouble srfvec[3];
  SpiceDouble phase = 0.0;
  SpiceDouble incdnc = 0.0;
  SpiceDouble emissn = 0.0;
  SpiceBoolean visibl = SPICEFALSE;
  SpiceBoolean lit = SPICEFALSE;

  illumf_c(
      method,
      target,
      ilusrc,
      (SpiceDouble)et,
      fixref,
      abcorr,
      obsrvr,
      spoint,
      &trgepc,
      srfvec,
      &phase,
      &incdnc,
      &emissn,
      &visibl,
      &lit);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outTrgepc) {
    *outTrgepc = (double)trgepc;
  }
  if (outSrfvec3) {
    outSrfvec3[0] = (double)srfvec[0];
    outSrfvec3[1] = (double)srfvec[1];
    outSrfvec3[2] = (double)srfvec[2];
  }
  if (outPhase) {
    *outPhase = (double)phase;
  }
  if (outIncdnc) {
    *outIncdnc = (double)incdnc;
  }
  if (outEmissn) {
    *outEmissn = (double)emissn;
  }
  if (outVisibl) {
    *outVisibl = visibl == SPICETRUE ? 1 : 0;
  }
  if (outLit) {
    *outLit = lit == SPICETRUE ? 1 : 0;
  }

  return 0;
}

int tspice_nvc2pl(
    const double *normal3,
    double konst,
    double *outPlane4,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outPlane4) {
    outPlane4[0] = 0.0;
    outPlane4[1] = 0.0;
    outPlane4[2] = 0.0;
    outPlane4[3] = 0.0;
  }

  SpiceDouble normal[3];
  normal[0] = normal3 ? (SpiceDouble)normal3[0] : 0.0;
  normal[1] = normal3 ? (SpiceDouble)normal3[1] : 0.0;
  normal[2] = normal3 ? (SpiceDouble)normal3[2] : 0.0;

  SpicePlane plane;
  nvc2pl_c(normal, (SpiceDouble)konst, &plane);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outPlane4) {
    outPlane4[0] = (double)plane.normal[0];
    outPlane4[1] = (double)plane.normal[1];
    outPlane4[2] = (double)plane.normal[2];
    outPlane4[3] = (double)plane.constant;
  }

  return 0;
}

int tspice_pl2nvc(
    const double *plane4,
    double *outNormal3,
    double *outKonst,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outNormal3) {
    outNormal3[0] = 0.0;
    outNormal3[1] = 0.0;
    outNormal3[2] = 0.0;
  }
  if (outKonst) {
    *outKonst = 0.0;
  }

  SpicePlane plane;
  plane.normal[0] = plane4 ? (SpiceDouble)plane4[0] : 0.0;
  plane.normal[1] = plane4 ? (SpiceDouble)plane4[1] : 0.0;
  plane.normal[2] = plane4 ? (SpiceDouble)plane4[2] : 0.0;
  plane.constant = plane4 ? (SpiceDouble)plane4[3] : 0.0;

  SpiceDouble normal[3];
  SpiceDouble outK = 0.0;

  pl2nvc_c(&plane, normal, &outK);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outNormal3) {
    outNormal3[0] = (double)normal[0];
    outNormal3[1] = (double)normal[1];
    outNormal3[2] = (double)normal[2];
  }
  if (outKonst) {
    *outKonst = (double)outK;
  }

  return 0;
}

int tspice_occult(
    const char *targ1,
    const char *shape1,
    const char *frame1,
    const char *targ2,
    const char *shape2,
    const char *frame2,
    const char *abcorr,
    const char *observer,
    double et,
    int *outOcltid,
    char *err,
    int errMaxBytes) {
  tspice_init_cspice_error_handling_once();

  if (errMaxBytes > 0) {
    err[0] = '\0';
  }
  if (outOcltid) {
    *outOcltid = 0;
  }

  SpiceInt ocltid = 0;
  occult_c(targ1, shape1, frame1, targ2, shape2, frame2, abcorr, observer, (SpiceDouble)et, &ocltid);
  if (failed_c()) {
    tspice_get_spice_error_message_and_reset(err, errMaxBytes);
    return 1;
  }

  if (outOcltid) {
    *outOcltid = (int)ocltid;
  }
  return 0;
}

File: packages/backend-node/src/domains/geometry.ts
import type {
  Found,
  GeometryApi,
  IllumfResult,
  IllumgResult,
  IluminResult,
  Pl2nvcResult,
  SpicePlane,
  SpiceVector3,
  SubPointResult,
} from "@rybosome/tspice-backend-contract";
import { invariant } from "@rybosome/tspice-core";

import type { NativeAddon } from "../runtime/addon.js";

export function createGeometryApi(native: NativeAddon): GeometryApi {
  return {
    subpnt: (method, target, et, fixref, abcorr, observer) => {
      const out = native.subpnt(method, target, et, fixref, abcorr, observer);
      invariant(out && typeof out === "object", "Expected subpnt() to return an object");
      invariant(
        Array.isArray(out.spoint) && out.spoint.length === 3,
        "Expected subpnt().spoint to be a length-3 array",
      );
      invariant(typeof out.trgepc === "number", "Expected subpnt().trgepc to be a number");
      invariant(
        Array.isArray(out.srfvec) && out.srfvec.length === 3,
        "Expected subpnt().srfvec to be a length-3 array",
      );
      return {
        spoint: out.spoint as SpiceVector3,
        trgepc: out.trgepc,
        srfvec: out.srfvec as SpiceVector3,
      } satisfies SubPointResult;
    },

    subslr: (method, target, et, fixref, abcorr, observer) => {
      const out = native.subslr(method, target, et, fixref, abcorr, observer);
      invariant(out && typeof out === "object", "Expected subslr() to return an object");
      invariant(
        Array.isArray(out.spoint) && out.spoint.length === 3,
        "Expected subslr().spoint to be a length-3 array",
      );
      invariant(typeof out.trgepc === "number", "Expected subslr().trgepc to be a number");
      invariant(
        Array.isArray(out.srfvec) && out.srfvec.length === 3,
        "Expected subslr().srfvec to be a length-3 array",
      );
      return {
        spoint: out.spoint as SpiceVector3,
        trgepc: out.trgepc,
        srfvec: out.srfvec as SpiceVector3,
      } satisfies SubPointResult;
    },

    sincpt: (method, target, et, fixref, abcorr, observer, dref, dvec) => {
      const out = native.sincpt(method, target, et, fixref, abcorr, observer, dref, dvec);
      if (!out.found) {
        return { found: false };
      }

      invariant(out && typeof out === "object", "Expected sincpt() to return an object");
      invariant(
        Array.isArray(out.spoint) && out.spoint.length === 3,
        "Expected sincpt().spoint to be a length-3 array",
      );
      invariant(typeof out.trgepc === "number", "Expected sincpt().trgepc to be a number");
      invariant(
        Array.isArray(out.srfvec) && out.srfvec.length === 3,
        "Expected sincpt().srfvec to be a length-3 array",
      );
      return {
        found: true,
        spoint: out.spoint as SpiceVector3,
        trgepc: out.trgepc,
        srfvec: out.srfvec as SpiceVector3,
      } satisfies Found<SubPointResult>;
    },

    ilumin: (method, target, et, fixref, abcorr, observer, spoint) => {
      const out = native.ilumin(method, target, et, fixref, abcorr, observer, spoint);
      invariant(out && typeof out === "object", "Expected ilumin() to return an object");
      invariant(typeof out.trgepc === "number", "Expected ilumin().trgepc to be a number");
      invariant(
        Array.isArray(out.srfvec) && out.srfvec.length === 3,
        "Expected ilumin().srfvec to be a length-3 array",
      );
      invariant(typeof out.phase === "number", "Expected ilumin().phase to be a number");
      invariant(typeof out.incdnc === "number", "Expected ilumin().incdnc to be a number");
      invariant(typeof out.emissn === "number", "Expected ilumin().emissn to be a number");
      return {
        trgepc: out.trgepc,
        srfvec: out.srfvec as SpiceVector3,
        phase: out.phase,
        incdnc: out.incdnc,
        emissn: out.emissn,
      } satisfies IluminResult;
    },

    illumg: (method, target, ilusrc, et, fixref, abcorr, observer, spoint) => {
      const out = native.illumg(method, target, ilusrc, et, fixref, abcorr, observer, spoint);
      invariant(out && typeof out === "object", "Expected illumg() to return an object");
      invariant(typeof out.trgepc === "number", "Expected illumg().trgepc to be a number");
      invariant(
        Array.isArray(out.srfvec) && out.srfvec.length === 3,
        "Expected illumg().srfvec to be a length-3 array",
      );
      invariant(typeof out.phase === "number", "Expected illumg().phase to be a number");
      invariant(typeof out.incdnc === "number", "Expected illumg().incdnc to be a number");
      invariant(typeof out.emissn === "number", "Expected illumg().emissn to be a number");

      return {
        trgepc: out.trgepc,
        srfvec: out.srfvec as SpiceVector3,
        phase: out.phase,
        incdnc: out.incdnc,
        emissn: out.emissn,
      } satisfies IllumgResult;
    },

    illumf: (method, target, ilusrc, et, fixref, abcorr, observer, spoint) => {
      const out = native.illumf(method, target, ilusrc, et, fixref, abcorr, observer, spoint);
      invariant(out && typeof out === "object", "Expected illumf() to return an object");
      invariant(typeof out.trgepc === "number", "Expected illumf().trgepc to be a number");
      invariant(
        Array.isArray(out.srfvec) && out.srfvec.length === 3,
        "Expected illumf().srfvec to be a length-3 array",
      );
      invariant(typeof out.phase === "number", "Expected illumf().phase to be a number");
      invariant(typeof out.incdnc === "number", "Expected illumf().incdnc to be a number");
      invariant(typeof out.emissn === "number", "Expected illumf().emissn to be a number");
      invariant(typeof out.visibl === "boolean", "Expected illumf().visibl to be a boolean");
      invariant(typeof out.lit === "boolean", "Expected illumf().lit to be a boolean");

      return {
        trgepc: out.trgepc,
        srfvec: out.srfvec as SpiceVector3,
        phase: out.phase,
        incdnc: out.incdnc,
        emissn: out.emissn,
        visibl: out.visibl,
        lit: out.lit,
      } satisfies IllumfResult;
    },

    occult: (targ1, shape1, frame1, targ2, shape2, frame2, abcorr, observer, et) => {
      const out = native.occult(targ1, shape1, frame1, targ2, shape2, frame2, abcorr, observer, et);
      invariant(typeof out === "number", "Expected occult() to return a number");
      return out;
    },

    nvc2pl: (normal, konst) => {
      const out = native.nvc2pl(normal, konst);
      invariant(Array.isArray(out) && out.length === 4, "Expected nvc2pl() to return a length-4 array");
      return out as unknown as SpicePlane;
    },

    pl2nvc: (plane) => {
      const out = native.pl2nvc(plane);
      invariant(out && typeof out === "object", "Expected pl2nvc() to return an object");
      invariant(
        Array.isArray(out.normal) && out.normal.length === 3,
        "Expected pl2nvc().normal to be a length-3 array",
      );
      invariant(typeof out.konst === "number", "Expected pl2nvc().konst to be a number");
      return {
        normal: out.normal as SpiceVector3,
        konst: out.konst,
      } satisfies Pl2nvcResult;
    },
  };
}
